marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/hub-entities.xqy

```diff
@@ -523,3 +525,47 @@ declare function hent:json-schema-generate($entity-title as xs:string, $uber-mod
   )
   return xdmp:to-json($uber-model)
 };
+
+(:
+  this function finds the first level facetable entityType properties and constrcuts and adds the rangeIndex array to
+  the entityModel. All the structured properties are ignored for now even if a property is modeled as facetable as per
+  https://project.marklogic.com/jira/browse/DHFPROD-5018
+:)
+declare %private function hent:add-indexes-for-entity-properties($entities as json:array) {
+  let $models := json:array-values($entities) ! xdmp:to-json(.)/object-node()
+  let $updated-models := json:array()
+
+  let $_ :=
+    for $model as map:map in $models
+      let $entity-type := map:get($model, "definitions")=>map:get(map:get($model, "info")=>map:get("title"))
+      let $entity-type-properties :=
+        let $empty-map := map:map()
+          return
+            if (fn:empty($entity-type)) then
+              let $_ := xdmp:log("Could not find entity definition with name: " || map:get($model, "info")=>map:get("title"))
+              return $empty-map
+            else
+              let $_ :=
+                if (fn:empty(map:get($entity-type, "rangeIndex"))) then map:put($entity-type, "rangeIndex", json:array())
+                else ()
+              return map:get($entity-type, "properties")
+
+      let $_ :=
+        for $entity-type-property in map:keys($entity-type-properties)
+          let $ref := map:get($entity-type-properties, $entity-type-property)=>map:get("$ref")
+          let $is-facetable :=
+            if (fn:empty($ref)) then
+              let $result :=
+                let $items := map:get($entity-type-properties, $entity-type-property)=>map:get("items")
+                return if(fn:empty($items)) then map:get($entity-type-properties, $entity-type-property)=>map:get("facetable")
+                else fn:not(fn:starts-with(map:get($items, "$ref"), "#")) and map:get($entity-type-properties, $entity-type-property)=>map:get("facetable")
+              return $result
+            else
+              let $items := map:get($entity-type-properties, $entity-type-property)=>map:get("items")
```

Comment:
I couldn't figure out what this "else" block was doing. I removed it locally and then ran the updated tests in this PR, and they still passed. If the top-level property has a "$ref", then it's a structured property, in which case we aren't yet supporting facetable on its child properties, right? 

This was the impl of the for loop that I tested locally (I like using "where" to avoid empty "else" statements, also think it makes the code more self-documenting):

```
(: Iterate over each top-level property and if it's facetable, add it to the rangeIndex array:)
    let $_ :=
      for $entity-type-property in map:keys($entity-type-properties)
      let $ref := map:get($entity-type-properties, $entity-type-property)=>map:get("$ref")
      where fn:empty($ref)
      return 
        let $is-facetable :=
          let $items := map:get($entity-type-properties, $entity-type-property)=>map:get("items")
          return 
            if (fn:empty($items)) then 
              map:get($entity-type-properties, $entity-type-property)=>map:get("facetable")
            else 
              fn:not(fn:starts-with(map:get($items, "$ref"), "#")) and 
              map:get($entity-type-properties, $entity-type-property)=>map:get("facetable")
        where $is-facetable
        return json:array-push(map:get($entity-type, "rangeIndex"), $entity-type-property)
```

Reply:
Addressed them in the latest commit